clang/test/SemaSYCL/union-kernel-param1.cpp

```diff
@@ -0,0 +1,173 @@
+// RUN: %clang_cc1 -I %S/Inputs -fsycl -fsycl-is-device -ast-dump %s | FileCheck %s
+
+// This test checks that compiler generates correct kernel arguments for
+// unions containing Arrays.
+
+#include <sycl.hpp>
+
+using namespace cl::sycl;
+
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void a_kernel(Func kernelFunc) {
+  kernelFunc();
+}
+
+template <typename T>
+union S {
+  T a[3];
+};
+
+int main() {
+
+  union union_acc_t {
+    int member_acc[2];
+  } union_acc;
+
+  S<int> s;
+
+  union foo_inner {
+    int foo_inner_x;
+    int foo_inner_y;
+    int foo_inner_z[2];
+  };
+
+  union foo {
+    int foo_a;
+    foo_inner foo_b[2];
+    int foo_c;
+  };
+
+  foo union_array[2];
+
+  a_kernel<class kernel_A>(
+      [=]() {
+        int array = union_acc.member_acc[1];
+      });
+
+  a_kernel<class kernel_B>(
+      [=]() {
+        foo local = union_array[1];
+      });
+
+  a_kernel<class kernel_C>(
+      [=]() {
+        int local = s.a[2];
+      });
+}
+
+// Check kernel_A parameters
+// CHECK: FunctionDecl {{.*}}kernel_A{{.*}} 'void (union union_acc_t, int, int)'
+// CHECK-NEXT: ParmVarDecl {{.*}} used _arg_ 'union union_acc_t':'union_acc_t'
+// CHECK-NEXT: ParmVarDecl {{.*}} used _arg_member_acc 'int'
```

Comment:
It looks decomposition is happening ? member_acc is the union array. This should not be passed as individual parameters right?  It looks like you are passing the 'not-decomposed' union as well as it's fields?